### ISSUE TITLE
pytest using napari crashes

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari-plugin-devtools#15]
Hi napari plugin devtools folks,

I'm trying to write tests for my [napari plugin](https://github.com/clEsperanto/napari_pyclesperanto_assistant) and [such a test](https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/7/files#diff-1ebd085e84798469b8eeeefa25807dd00ba36341c85e3c0788c16bdcf5c3f97cR9):
```
import napari
viewer = napari.Viewer(show=False)
```
[fails in the github CI](https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/7/checks?check_run_id=2290288069#step:10:20) with this error:
```
Error: Process completed with exit code 134.
```

I've also tried with make_napari_viewer fixtures as done in [napari tests](https://github.com/napari/napari/blob/master/napari/_tests/test_adding_removing.py#L11), but the error is persistend. I assume I presume I'm [calling tests](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/4112588158e0389975ecdfe2c50f818e9cb0e77f/.github/workflows/test_and_deploy.yml#L89) in a wrong way...
Is there somewhere an example test showing us plugin developers how to write a test involving the napari GUI? 

Any hint is appreciated!

Thanks,
Robert

**Comment by tlambert03** [comment]
Hey @haesleinhuepf, I would definitely expect that you would _either_ need to use `make_napari_viewer` or use `qtbot` (or just `qapp`) from pytest-qt directly.  you get exactly the same error when you use `make_napari_viewer`?  do you have an example of that?

> Is there somewhere an example test showing us plugin developers how to write a test involving the napari GUI

our docs on testing with a viewer are here:  https://napari.org/developers/testing.html#writing-tests

**Comment by haesleinhuepf** [comment]
Hey @tlambert03 ,

thanks for the link! A little confusion stays: Do I have to define the `make_test_napari` myself (that's how I did so far) or is there a way of importing it from napari?



> do you have an example of that?

Yes, here you go. Code:
https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/8/files#diff-1ebd085e84798469b8eeeefa25807dd00ba36341c85e3c0788c16bdcf5c3f97cR28

Error:
https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/8/checks?check_run_id=2290892049#step:10:21

Again, I think there is some how-to-run-the-tests thingy missing...

Thanks for your support!

**Comment by haesleinhuepf** [comment]
Btw. I found multiple reports online  ([example](https://stackoverflow.com/questions/48290403/process-finished-with-exit-code-134-interrupted-by-signal-6-sigabrt)) suggesting that this error is an out-of-memory error

**Comment by tlambert03** [comment]
> A little confusion stays: Do I have to define the make_test_napari myself (that's how I did so far) or is there a way of importing it from napari?

yeah, I don't blame you for being a little confused... this is one of the more confusing aspects of pytest (it's not a napari thing at all).  But the key is in that first hint in the napari docs:

> note: fixtures in pytest can be a little mysterious, since it’s not always clear where they are coming from. In this case, using a pytest-qt fixture looks like this:
>
>  just by putting `qtbot` in the list of arguments...

There's important information to be found in the linked docs about [`pytest-fixtures`](https://docs.pytest.org/en/stable/fixture.html) ... but the main point is this: to use a pytest fixture, all you do is put the name of the fixture in the function arguments.  So, literally, all you should do is what is shown in step 2 in the [testing docs](https://napari.org/developers/testing.html#writing-tests)

```python
def test_whatever2(make_napari_viewer):
    viewer = make_napari_viewer()
    # don't call viewer.close...
    # don't use gui_qt...
```
and as long as napari is in your environment, pytest will find that fixture (in this regard napari acts as a pytest "plugin").  
so, basically, _don't_ do what you did where you redefined `make_test_viewer` in your code there ... That's just how pytest works :/  we all love to hate it :joy:

all that said, I'm able to pull and run your current tests successfully locally, are you?  If so, then we need to focus more on what's going on in CI

**Comment by haesleinhuepf** [comment]
> all that said, I'm able to pull and run your current tests successfully locally, are you? If so, then we need to focus more on what's going on in CI

Correct, locally all is fine.

I just [cleaned up as you suggested](https://github.com/clEsperanto/napari_pyclesperanto_assistant/pull/8/commits/4e0c917fe243523f62e64a561ce9585d272500fd) but now already locally there is an issue:

> and as long as napari is in your environment, pytest will find that fixture (in this regard napari acts as a pytest "plugin").

Locally, I get this error when calling pytest:
```
  def test_complex_workflow(make_test_viewer):
E       fixture 'make_test_viewer' not found
```

So, the fixture must be listed somewhere in the config, I guess

**Comment by tlambert03** [comment]
it's called `make_napari_viewer` ([step 2 here](https://napari.org/developers/testing.html#writing-tests))

**Comment by tlambert03** [comment]
note: if you use that, you may start getting things like "ERROR at teardown of test_complex_workflow" ... (at least I get them when testing your plugin).  That means that some of your Qt code is not cleaning up appropriately, and you're leaving widgets around.  This is an important, but unfortunately trickier part of using Qt.  Usually it can be solved by making sure that every widget you create is given a parent widget ... but that's an oversimplification of the solution.  It's a good thing to solve, but if you want to punt on solving that, you might want to avoid `make_napari_viewer` for now and just use `qtbot` directly

**Comment by haesleinhuepf** [comment]
> it's called make_napari_viewer

OMG I should code less and read more. Thanks!

Ok, I now receive an Assertion Error: "Widgets leaked." maybe related to https://github.com/napari/napari/issues/2089 , maybe not

> Usually it can be solved by making sure that every widget you create is given a parent widget ...

Ok, I will dive into this myself.

> It's a good thing to solve, but if you want to punt on solving that, you might want to avoid make_napari_viewer for now and just use qtbot directly

Or do this :D 

Thanks again for the support! 🤝

**Comment by tlambert03** [comment]
> Ok, I now receive an Assertion Error: "Widgets leaked."

yep, that's the part I was mentioning above.  I'm afraid that's on your end :joy:


i see that it's still failing on CI.  this is probably going to be a guess and check sort of thing.  If I were you I'd simplify and try a generally simpler environment, simpler pytest command (maybe leave out pytest-benchmark, coverage etc...)  we routinely use `make_napari_viewer` on CI, so it has to be something about your test setup there, but nothing immediately jumps out at me

**Comment by haesleinhuepf** [comment]
> I'd simplify and try a generally simpler environment

My environment is a mixture of the environment you setup for pyclesperanto (awesome CI work) and what the cookiecutter provided. It took me a while to mix these two properly :-)

**Comment by tlambert03** [comment]
actually... I think i figured out what's causing your Abort error:  you don't have pytest-xvfb anywhere in your test dependencies.  See this comment your github yaml:

>     these libraries, along with pytest-xvfb (added in the `deps` in tox.ini),
>     enable testing on Qt on linux

In the cookiecutter, this dependency is provided as a test-only dependency [in the tox file:](https://github.com/napari/cookiecutter-napari-plugin/blob/b89b0524037227b113e6aa512354afb698400e3f/%7B%7Bcookiecutter.plugin_name%7D%7D/tox.ini#L31)
... but here you're not using the tox file, just calling pytest directly.  try adding `pip install pytest-xvfb` to the linux section:

```yaml
      # these libraries, along with pytest-xvfb (added in the `deps` in tox.ini),
      # enable testing on Qt on linux
      - name: Install Linux libraries
        if: runner.os == 'Linux'
        run: |
          pip install pytest-xvfb
          sudo apt-get install -y libdbus-1-3 libxkbcommon-x11-0 libxcb-icccm4 \
            libxcb-image0 libxcb-keysyms1 libxcb-randr0 libxcb-render-util0 \
            libxcb-xinerama0 libxcb-xinput0 libxcb-xfixes0
```
